tutorials/Tutorial MIS & QAOA.ipynb

```diff
@@ -0,0 +1,394 @@
+    "    plt.figure(figsize=(12,6))\n",
+    "    plt.xlabel(\"bitstrings\")\n",
+    "    plt.ylabel(\"counts\")\n",
+    "    plt.bar(C.keys(), C.values(), width=0.5, color = color_dict.values())\n",
+    "    plt.xticks(rotation='vertical')\n",
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plot_distribution(example_dict)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The bitstrings '01011' and '00111' (in red) correspond to the two MIS of the graph. The goal of QAOA is to choregraph interferences between the basis states, in order to maximize the frequency of the MIS states. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Optimization "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We estimate the cost of a sampled state vector by making an average over the samples. Determining the cost of a given bitstring takes polynomial time. The average estimate is then used in the classical loop to optimize the variational parameters $\\tau$ and $t$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_cost_colouring(z,G,penalty=10):\n",
+    "    \"\"\"G: the graph (igraph)\n",
+    "       z: a binary colouring\n",
+    "       returns the cost of the colouring z, depending on the adjacency of the graph\"\"\"\n",
+    "    cost = 0 \n",
+    "    A = G.get_adjacency()\n",
+    "    z = np.array(tuple(z),dtype=int)\n",
+    "    for i in range(len(z)):\n",
+    "        for j in range(i,len(z)):\n",
+    "            cost += A[i][j]*z[i]*z[j]*penalty # if there's an edge between i,j and they are both in R state.\n",
+    "    \n",
+    "    cost -= np.sum(z) #to count for the 0s instead of the 1s \n",
+    "    return cost \n",
+    "\n",
+    "def get_cost(counter,G):\n",
+    "    cost = 0 \n",
+    "    for key in counter.keys():\n",
+    "        cost_col = get_cost_colouring(key,G)\n",
+    "        cost += cost_col * counter[key]\n",
+    "    return cost / sum(counter.values())"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "get_cost_colouring('00111',G)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def func(param,*args):\n",
+    "    G = args[0]\n",
+    "    C = quantum_loop(param, r=reg)\n",
+    "    cost = get_cost(C,G)\n",
+    "    return cost\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# QAOA for p = 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`minimize` takes the as arguments `func`, the graph `G`and an initial `x0` point for the simplex."
```

Comment:
Maybe a line or two giving some remarks about the optimization function and method used would be nice.